protocols/kad/src/behaviour.rs

```diff
@@ -836,229 +922,232 @@ where
+
+    /// The result of [`Kademlia::get_record`].
+    GetRecordResult(GetRecordResult),
+
+    /// The result of [`Kademlia::put_record`].
+    PutRecordResult(PutRecordResult),
+
+    /// A new peer in the DHT has been discovered during an iterative query.
     Discovered {
-        /// Id of the node that was discovered.
+        /// The identifier of the discovered peer.
         peer_id: PeerId,
-        /// Addresses of the node.
+        /// The known addresses of the discovered peer.
         addresses: Vec<Multiaddr>,
-        /// How the reporter is connected to the reported.
+        /// The connection status of the reported peer, as seen by the local
+        /// peer.
         ty: KadConnectionType,
     },
 
-    /// A node has been added to a k-bucket.
-    KBucketAdded {
-        /// Id of the node that was added.
-        peer_id: PeerId,
-        /// If `Some`, this addition replaced the value that is inside the option.
-        replaced: Option<PeerId>,
+    /// A peer in the DHT has been added to the routing table.
+    RoutingUpdated {
+        /// The ID of the peer that was added to a bucket in the routing table.
+        new_peer: PeerId,
+        /// The ID of the peer that was evicted from the routing table to make
+        /// room for the new peer, if any.
+        old_peer: Option<PeerId>,
     },
+}
 
-    /// Result of a `FIND_NODE` iterative query.
-    FindNodeResult {
-        /// The key that we looked for in the query.
-        key: PeerId,
-        /// List of peers ordered from closest to furthest away.
-        closer_peers: Vec<PeerId>,
-    },
+/// The result of [`Kademlia::get_record`].
+pub type GetRecordResult = Result<GetRecordOk, GetRecordError>;
+
+/// The successful result of [`Kademlia::get_record`].
+#[derive(Debug, Clone)]
+pub struct GetRecordOk {
+    pub records: Vec<Record>
+}
+
+/// The error result of [`Kademlia::get_record`].
+#[derive(Debug, Clone)]
+pub enum GetRecordError {
+    NotFound { key: Multihash, closest_peers: Vec<PeerId> },
+    QuorumFailed { key: Multihash, records: Vec<Record>, quorum: usize },
+    Timeout { key: Multihash, records: Vec<Record>, quorum: usize }
+}
+
+/// The result of [`Kademlia::put_record`].
+pub type PutRecordResult = Result<PutRecordOk, PutRecordError>;
+
+/// The successful result of [`Kademlia::put_record`].
+#[derive(Debug, Clone)]
+pub struct PutRecordOk {
+    pub key: Multihash
+}
+
+/// The error result of [`Kademlia::put_record`].
+#[derive(Debug, Clone)]
+pub enum PutRecordError {
+    QuorumFailed { key: Multihash, num_results: usize, quorum: usize },
+    Timeout { key: Multihash, num_results: usize, quorum: usize },
+    LocalStorageError(RecordStorageError)
+}
+
+/// The result of [`Kademlia::bootstrap`].
+pub type BootstrapResult = Result<BootstrapOk, BootstrapError>;
+
+/// The successful result of [`Kademlia::bootstrap`].
+#[derive(Debug, Clone)]
+pub struct BootstrapOk {
+    pub peer: PeerId
```

Comment:
What is that value?

Reply:
The peer ID that is the target of the bootstrapping query - that is, either the local peer ID (first query) or a random peer ID for a specific bucket (follow-up queries). I can add a doc-comment.

Reply:
To me it sounded obvious that `BootstrapOk` was only produced once at the very end of the bootstrapping process.

Reply:
Which, to me, shows that more documentation is needed here.

Reply:
How about the [existing documentation of `Kademlia::bootstrap`](https://github.com/libp2p/rust-libp2p/blob/ed07e679c8af9b2c298b3e4afcd066fb22e026c6/protocols/kad/src/behaviour.rs#L276-L290)? I think it does explain it in sufficient detail, but let me know what you think is missing.